drivers/tee/optee/optee.c

```diff
@@ -177,9 +178,189 @@ static void u64_to_regs(uint64_t val, uint32_t *reg0, uint32_t *reg1)
+	arg->params[0].u.tmem.size = shm->size;
+	arg->params[0].u.tmem.shm_ref = (uint64_t)shm;
+	arg->ret = TEEC_SUCCESS;
+	return;
+out:
+	tee_shm_free(dev, shm);
+}
+
+static void handle_cmd_free(const struct device *dev, struct optee_msg_arg *arg)
+{
+	int rc;
+
+	if (!check_param_input(arg)) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	/*
+	 * TODO: OPTEE_SHM_TYPE_APPL type should be handled
+	 */
+	if (arg->params[0].u.value.a != OPTEE_RPC_SHM_TYPE_KERNEL) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	rc = tee_rm_shm(dev, (struct tee_shm *)arg->params[0].u.value.b);
+	if (rc) {
+		arg->ret = TEEC_ERROR_OUT_OF_MEMORY;
+		return;
+	}
+}
+
+static void handle_cmd_get_time(const struct device *dev, struct optee_msg_arg *arg)
+{
+	int64_t ticks;
+	int64_t up_secs;
+	int64_t up_nsecs;
+
+	if (arg->num_params != 1 ||
+	    (arg->params[0].attr & OPTEE_MSG_ATTR_TYPE_MASK)
+	    != OPTEE_MSG_ATTR_TYPE_VALUE_OUTPUT) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	ticks = k_uptime_ticks();
+
+	up_secs = ticks / CONFIG_SYS_CLOCK_TICKS_PER_SEC;
+	up_nsecs = k_ticks_to_ns_floor64(ticks - up_secs * CONFIG_SYS_CLOCK_TICKS_PER_SEC);
+	arg->params[0].u.value.a = up_secs;
+	arg->params[0].u.value.b = up_nsecs;
+
+	arg->ret = TEEC_SUCCESS;
+}
+
+static void handle_cmd_notify(const struct device *dev, struct optee_msg_arg *arg)
+{
+	/*
+	 * TODO: Implement wait for completion based on zephyr workq
+	 * Just sleep for 10 msec for now.
+	 */
+	k_sleep(K_MSEC(10));
+	arg->ret = TEEC_SUCCESS;
+}
+
+static void handle_cmd_wait(const struct device *dev, struct optee_msg_arg *arg)
+{
+	if (check_param_input(arg)) {
+		arg->ret = TEEC_ERROR_BAD_PARAMETERS;
+		return;
+	}
+
+	k_sleep(K_MSEC(arg->params[0].u.value.a));
+
+	arg->ret = TEEC_SUCCESS;
+}
+
+static void free_shm_pages(void **pages)
+{
+	/*
+	 * Clean allocated pages if needed. Some function calls requires pages
+	 * allocation which should be freed after processing new request.
+	 * It is safe to free this list when another SHM op (e,g. another alloc
+	 * or free) was received.
+	 */
+	if (*pages) {
+		k_free(*pages);
+		*pages = NULL;
+	}
+}
+
+static uint32_t handle_func_rpc_call(const struct device *dev, struct tee_shm *shm,
+				     void **pages)
+{
+	struct optee_msg_arg *arg = shm->addr;
+
+	switch (arg->cmd) {
+	case OPTEE_RPC_CMD_SHM_ALLOC:
+		free_shm_pages(pages);
+		handle_cmd_alloc(dev, arg, pages);
+		break;
+	case OPTEE_RPC_CMD_SHM_FREE:
+		handle_cmd_free(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_GET_TIME:
+		handle_cmd_get_time(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_NOTIFICATION:
+		handle_cmd_notify(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_SUSPEND:
+		handle_cmd_wait(dev, arg);
+		break;
+	case OPTEE_RPC_CMD_I2C_TRANSFER:
+		/* TODO: i2c transfer case is not implemented right now */
+		return TEEC_ERROR_NOT_IMPLEMENTED;
+	default:
+		/* TODO: supplicant commands are still not implemented */
+		/* handle_rpc_supp_cmd(ctx, optee, arg); */
```

Comment:
No commented out code, please.  Have you ran checkpatch, BTW?

Reply:
```
Rule 2.2 | Required | There shall be no dead code
```

Reply:
fixed